docs/platform-branches.md

```diff
@@ -0,0 +1,98 @@
+<!--
+     Copyright 2024, Proofcraft Pty Ltd
+     SPDX-License-Identifier: CC-BY-SA-4.0
+-->
+
+# Automated Platform Branches
+
+The following branches for additional platforms are currently automatically
+rebased and tested on each push to `master`:
+
+- [imx8mm-fpu-ver](https://github.com/seL4/l4v/tree/imx8-fpu-ver)
+- [exynos5-ver](https://github.com/seL4/l4v/tree/exynos5-ver)
+
+## How this works
+
+The `L4V_PLAT` configuration setting provides an additional dimension to the
+configuration space for verified platforms in addition to `L4V_ARCH` and
+`L4V_FEATURES`. To test an `L4V_ARCH`, `L4V_FEATURES`, and `L4V_PLAT`
+combination, the test setup will attempt to load a corresponding kernel config
+from the seL4 repository under
+
+```sh
+configs/${L4V_ARCH}_${L4V_PLAT}_${L4V_FEATURES}_verified.cmake
+```
+
+e.g. `ARM_HYP_exynos5_verified.cmake`. Tests on the master branch will currently
+keep `L4V_PLAT` empty, and tests on the platform branches will provide the
+appropriate platform string.
+
+To keep platform branches up to date, there is [a GitHub action][rebase-action]
+that automatically attempts to rebase all platform branches over each new push
+to the master branch. If that rebase succeeds, this action will push the rebased
+platform branch to a candidate branch `<branch>-rebased`. The platform branch,
+and therefore the `-rebased` branch have a corresponding action that in turn is
+triggered by this push and runs a full proof test for this branch. If the test
+succeeds, the branch action replaces the original platform branch with the
+candidate `<branch>-rebased` branch by force push. If either step fails, CI will
+signal test failure and require manual intervention.
```

Comment:
who gets notified, or where can we see that it broke?

Reply:
Ooh, and to add to that, the above I was thinking of l4v master... but does the same thing happen with seL4 master changing (or going via manifest)?

Reply:
> Ooh, and to add to that, the above I was thinking of l4v master... but does the same thing happen with seL4 master changing (or going via manifest)?

Interesting question, I don't think we have thought through that scenario yet. The workflow triggers on seL4 changes, but there won't be anything to do for the rebase, since neither l4v master nor any of the l4v rebase branches will have changed. This means, it will not trigger a re-test of the platform branches. There could in theory be a change in seL4 that affects only those platforms and is invisible to the other ones, and those would be missed until the next l4v PR.

There is a `workflow_dispatch` trigger (= the one that is triggered by manifest/seL4 updates) in the branch workflows, but I'm not sure you can trigger anything else than the master branch with these. I should investigate.

Reply:
> Who gets notified, or where can we see that it broke? (either the rebase, or the regression test post-rebase)

Who gets notified is inscrutable GitHub magic. It does seem to figure out who pushed/merged/authored and notify that person. You can also see it in the test badges in the README -- if one of them is red and has a platform branch name, one of these tests has failed. If you then go to the Actions tab, you can see which ones.

I seem to get failure notifications on all actions that are not PRs from someone else. Not sure if that is because I'm watching the repository or if it is because the seL4 CI account re-routes to me and it does some of these pushes. 

> Further: if the rebase succeeds and the PR fails, will the next push to master result in another rebase attempt? 

Yes, it will trigger another attempt. If it failed at the rebase stage, it is likely to fail again if the last one failed.

> Will this get blocked/delayed by an existing regression test running on the rebased branch

A rebase failure won't interact with any running tests.

If two triggers made it past the rebase stage, the workflows will both keep running and they will both try to force-push their own state at the end. There is a race condition on which one wins, but only a working one will end up as the main platform branch. Unless the change makes something significantly faster, they'll proceed in the right order (i.e. later one wins if it passes).

If it's the wrong order, we can manually re-trigger the last one to run again, but I haven't seen that happening yet.

Reply:
> Right, because if we commit something to l4v master, that will trigger a rebase etc, and we will most likely do that at the same time as we advance the manifest to seL4 master... Do situations like bumping the seL4 version in cases where we expect the proof to go through trigger a normal verification test? I'd guess so, 

Yes, every seL4 change will trigger a re-run (even preprocess-equivalent ones these days)

> but then these ones won't trigger since there's nothing to rebase in that situation and they're not linked to the manifest. It wouldn't be the end of the world, but good to know/document, as I'm clearly a bit confused.

Yes, good idea. Will add.